### PR TITLE
i228: tested authService

### LINKED ISSUE #228: finish auth service tests
- [ ] test that sign streams close when sign in has finished
- [ ] test that sign streams close when sign in errors
- [ ] test sign out - dispatches a StoreProblem action on error

- [ ] test sign in with email
- [ ] test sign up account with email

### PR DESCRIPTION
**Pull request by tamari-gray**
fixes #228

### CODE REVIEW

**Review by tamari-gray — COMMENTED** [review]
think she gtg :)

File: lib/middleware/auth/sign_in_with_email.dart [hunk 1]
```diff
@@ -11,20 +9,9 @@ class SignInWithEmailMiddleware
       : super((store, action, next) async {
           next(action);
 
-          // set the UI to waiting
```
[the remark this answers is not in the document]
Comment (nickmeinhold) [reply lib/middleware/auth/sign_in_with_email.dart:14]: nice!

File: lib/middleware/auth/sign_out_user.dart [hunk 2]
```diff
@@ -9,9 +9,9 @@ class SignOutUserMiddleware extends TypedMiddleware<AppState, SignOutUser> {
           next(action);
```
[the remark this answers is not in the document]
Comment (nickmeinhold) [reply lib/middleware/auth/sign_out_user.dart:10]: great 👍

File: lib/middleware/auth/sign_up_with_email.dart [hunk 3]
```diff
@@ -11,20 +9,9 @@ class SignUpWithEmailMiddleware
       : super((store, action, next) async {
           next(action);
 
-          // set the UI to waiting
```
[the remark this answers is not in the document]
Comment (nickmeinhold) [reply lib/middleware/auth/sign_up_with_email.dart:14]: love it

File: lib/services/auth_service.dart [hunk 4]
```diff
@@ -142,16 +143,28 @@ class AuthService {
   ///  * `ERROR_USER_DISABLED` - If the user has been disabled (for example, in the Firebase console)
   ///  * `ERROR_TOO_MANY_REQUESTS` - If there was too many attempts to sign in as this user.
   ///  * `ERROR_OPERATION_NOT_ALLOWED` - Indicates that Email & Password accounts are not enabled.
-  Future<ReduxAction> signInWithEmail(String email, String password) async {
+  Stream<ReduxAction> emailSignInStream(String email, String password) async* {
```
[the remark this answers is not in the document]
Comment (nickmeinhold) [reply lib/services/auth_service.dart:146]: much nicer - thanks!

File: lib/services/auth_service.dart [hunk 5]
```diff
@@ -169,16 +182,30 @@ class AuthService {
   ///  * `ERROR_WEAK_PASSWORD` - If the password is not strong enough.
   ///  * `ERROR_INVALID_EMAIL` - If the email address is malformed.
   ///  * `ERROR_EMAIL_ALREADY_IN_USE` - If the email is already in use by a different account.
-  Future<ReduxAction> signUpWithEmail(String email, String password) async {
+  Stream<ReduxAction> emailSignUpStream(String email, String password) async* {
```
[the remark this answers is not in the document]
Comment (nickmeinhold) [reply lib/services/auth_service.dart:185]: boom!

File: lib/services/auth_service.dart [hunk 6]
```diff
@@ -188,7 +215,7 @@ class AuthService {
 
   Future<ReduxAction> signOut() async {
     try {
-      await FirebaseAuth.instance.signOut();
```
[the remark this answers is not in the document]
Comment (nickmeinhold) [reply lib/services/auth_service.dart:191]: Heck yeah! Thanks for making my code better!

File: test/mocks/auth/firebase_auth_mocks.dart [hunk 7]
```diff
@@ -13,13 +15,24 @@ class FakeFirebaseAuthPeriodic extends Fake implements FirebaseAuth {
       });
 }
 
-class FakeFirebaseAuthWithError extends Fake implements FirebaseAuth {
+class FakeFirebaseAuthSignInException extends Fake implements FirebaseAuth {
```
Comment (nickmeinhold) [inline test/mocks/auth/firebase_auth_mocks.dart:18]: The name here is very specific and so I don't think needs any documentation

File: test/mocks/auth/firebase_auth_mocks.dart [hunk 8]
```diff
@@ -13,13 +15,24 @@ class FakeFirebaseAuthPeriodic extends Fake implements FirebaseAuth {
...
   @override
-  Stream<auth.User> authStateChanges() => Stream.fromIterable([FakeAuthUser()]);
+  Future<UserCredential> signInWithCredential(AuthCredential credential) =>
+      throw FirebaseAuthException(
+          code: 'user-not-found', message: 'test error: cant find user');
+}
 
+class FakeFirebaseAuthThrows extends Fake implements FirebaseAuth {
```
Comment (nickmeinhold) [inline test/mocks/auth/firebase_auth_mocks.dart:25]: The name here is a bit more general so I think some commenting would be really helpful - is that OK?

File: test/mocks/auth/firebase_auth_mocks.dart [hunk 9]
```diff
@@ -13,13 +15,24 @@ class FakeFirebaseAuthPeriodic extends Fake implements FirebaseAuth {
...
+  @override
+  Future<void> signOut() => throw Exception('firebaseAuth.signOut');
+
+  @override
+  Future<auth.UserCredential> signInWithEmailAndPassword(
+          {String email, String password}) =>
+      throw Exception('firebaseAuth.signInWithEmailAndPassword');
 }
 
 class FakeFirebaseAuth1 extends Fake implements FirebaseAuth {
```
Comment (nickmeinhold) [inline test/mocks/auth/firebase_auth_mocks.dart:38]: I know this wasn't commented before (that's my bad) but would you mind adding a brief comment that indicates what the class is for and how/where it's used?

File: test/mocks/auth/google_signin_mocks.dart [hunk 10]
```diff
@@ -22,6 +24,11 @@ class FakeGoogleSignInThrows extends Fake implements GoogleSignIn {
   Future<GoogleSignInAccount> signIn() {
     throw Exception('GoogleSignIn.signIn');
   }
+
```
Comment (nickmeinhold) [inline test/mocks/auth/google_signin_mocks.dart:27]: could you add to the class comment what has changed with the new code?

File: test/mocks/mock_firebase_platform.dart [hunk 11]
```diff
@@ -0,0 +1,4 @@
+import 'package:firebase_core_platform_interface/firebase_core_platform_interface.dart';
+import 'package:mockito/mockito.dart';
+
+class MockFirebasePlatform extends Mock implements FirebasePlatform {}
```
Comment (nickmeinhold) [inline test/mocks/mock_firebase_platform.dart:4]: I think it'd be great if we both got into the habit of commenting the classes we make in the mocks files - could you add a comment here?

File: test/unit/middleware/auth/sign_in_with_email_middleware_test.dart [hunk 12]
```diff
@@ -5,66 +5,48 @@ import 'package:crowdleague/actions/navigation/remove_current_page.dart';
...
 import 'package:test/test.dart';
 
+import '../../../mocks/redux/redux_action_mocks.dart';
 import '../../../mocks/redux/redux_store_mocks.dart';
 import '../../../mocks/services/auth_service_mocks.dart';
 import '../../util/testing_utils.dart';
 
 void main() {
   group('SignInWithEmailMiddleware', () {
-    test('on successfull sign with firebase, dispatches correct actions',
```
[the remark this answers is not in the document]
Comment (nickmeinhold) [reply test/unit/middleware/auth/sign_in_with_email_middleware_test.dart:17]: awesome!

[the remark this answers is not in the document]
Comment (nickmeinhold) [reply test/unit/middleware/auth/sign_in_with_email_middleware_test.dart:17]: so much nice that way, thanks heaps!

File: test/unit/services/auth_service_test.dart [hunk 13]
```diff
@@ -1,22 +1,27 @@
...
+import 'package:mockito/mockito.dart';
 import 'package:test/test.dart';
 
 import '../../mocks/auth/apple_signin_mocks.dart';
 import '../../mocks/auth/firebase_auth_mocks.dart';
 import '../../mocks/auth/google_signin_mocks.dart';
-
-// TODO: test that sign streams close when sign in has finished
-// TODO: test that sign streams close when sign in errors
-// TODO: test sign out - dispatches a StoreProblem action on error
```
[the remark this answers is not in the document]
Comment (nickmeinhold) [reply test/unit/services/auth_service_test.dart:17]: you're a legend!  👋  TODOs... good riddance!

File: test/unit/services/auth_service_test.dart [hunk 14]
```diff
@@ -1,22 +1,27 @@
...
+import '../../mocks/mock_firebase_platform.dart';
 
 void main() {
+  // set up firebase
+  setUp(() {
+    Firebase.delegatePackingProperty = MockFirebasePlatform();
```
[the remark this answers is not in the document]
Comment (nickmeinhold) [reply test/unit/services/auth_service_test.dart:23]: Ooooooh nice!  Now I get what the `MockFirebasePlatform` was for!

File: test/unit/services/auth_service_test.dart [hunk 15]
```diff
@@ -147,5 +156,125 @@ void main() {
...
+          emitsInOrder(<dynamic>[
+            UpdateEmailAuthOptionsPage(step: AuthStep.signingInWithEmail),
+            UpdateEmailAuthOptionsPage(step: AuthStep.waitingForInput),
+            RemoveCurrentPage(),
+            emitsDone,
+          ]));
+    });
+
+    test(
+        'emailSignInStream catches firebaseAuthExceptions and emits StoreProblem actions',
```
Comment (nickmeinhold) [inline test/unit/services/auth_service_test.dart:176]: I think StoreProblem should be AddProblem?

File: test/unit/services/auth_service_test.dart [hunk 16]
```diff
@@ -147,5 +156,125 @@ void main() {
...
+                  'info',
+                  BuiltMap<String, Object>(
+                      {'code': 'test error: cant find user'}))
+              ..having((p) => p.problem.message, 'message',
+                  equals('firebase auth exception error')),
+            emitsDone
+          ]));
+    });
+
+    test('emailSignInStream catches errors and emits StoreProblem actions',
```
Comment (nickmeinhold) [inline test/unit/services/auth_service_test.dart:204]: StoreProblem -> AddProblem?

File: test/unit/services/auth_service_test.dart [hunk 17]
```diff
@@ -147,5 +156,125 @@ void main() {
...
+        () async {
+      // Init service that throws exception when signing in with firebase
+      final service = AuthService(FakeFirebaseAuthThrows(), null, null);
+
+      /// Check that emailSignInStream emits addProblem action with info from caught error.
```
Comment (nickmeinhold) [inline test/unit/services/auth_service_test.dart:209]: addProblem -> AddProblem (seems like a super nit but the meaning the case conveys here is worth caring about I reckon)

File: test/unit/services/auth_service_test.dart [hunk 18]
```diff
@@ -147,5 +156,125 @@ void main() {
...
+      final mockGoogleSignIn = MockGoogleSignIn();
+      final mockAppleSignIn = MockAppleSignIn();
+      final authService =
+          AuthService(mockFireBaseAuth, mockGoogleSignIn, mockAppleSignIn);
+
+      await authService.signOut();
+
+      verify(mockFireBaseAuth.signOut());
+      verify(mockGoogleSignIn.signOut());
+      // TODO: add apple sign in when #232 is completed. https://github.com/crowdleague/crowdleague/issues/232
```
Comment (nickmeinhold) [inline test/unit/services/auth_service_test.dart:237]: grab me for a chat about this if you don't mind - I think I'd like to get use away from allowing TODOs in the dev branch

File: test/unit/services/auth_service_test.dart [hunk 19]
```diff
@@ -147,5 +156,125 @@ void main() {
...
+        type: ProblemType.signOut,
+        traceString: '',
+      );
+
+      final error = await authService.signOut() as AddProblem;
+
+      expect(error.problem.type, testAddProblem.problem.type);
+      expect(error.problem.message, testAddProblem.problem.message);
+    });
+    test(
```
Comment (nickmeinhold) [inline test/unit/services/auth_service_test.dart:260]: super super nit - could you add a line b/w tests here? (whitespace)

**Review by nickmeinhold — CHANGES_REQUESTED** [review]
Awesome job mate!!! A few minor nits, just requesting comments in a few places really, code is really great - thanks heaps for making these improvements!

File: test/unit/services/auth_service_test.dart (hunk 18 above)
Comment (nickmeinhold) [reply test/unit/services/auth_service_test.dart:237]: I made [Style Guide > avoid-pushing-todos-to-dev](https://github.com/crowdleague/crowdleague/wiki/Style-Guide-(WIP)#avoid-pushing-todos-to-dev) and will add a note to discuss on Monday 🙂

Comment (nickmeinhold) [reply test/unit/services/auth_service_test.dart:237]: I've added an item to Monday standup

**Review by nickmeinhold — APPROVED** [review]
Really great work mate, thank you!
